File: frontend/src/lib/easing/easing.ts
export type EasingName =
  | "linear"
  | "easeOutCubic"
  | "easeInOutCubic"
  | "easeInOutSine";

function clamp01(t: number): number {
  if (Number.isNaN(t)) return 0;
  if (t < 0) return 0;
  if (t > 1) return 1;
  return t;
}

export function linear(t: number): number {
  return clamp01(t);
}

export function easeOutCubic(t: number): number {
  const x = clamp01(t);
  const inv = 1 - x;
  return 1 - inv * inv * inv;
}

export function easeInOutCubic(t: number): number {
  const x = clamp01(t);
  if (x < 0.5) {
    return 4 * x * x * x;
  }
  const inv = -2 * x + 2;
  return 1 - (inv * inv * inv) / 2;
}

export function easeInOutSine(t: number): number {
  const x = clamp01(t);
  return -(Math.cos(Math.PI * x) - 1) / 2;
}

export const EASING_FUNCTIONS: Record<EasingName, (t: number) => number> = {
  linear,
  easeOutCubic,
  easeInOutCubic,
  easeInOutSine,
};

export function interpolate(
  start: number,
  end: number,
  t: number,
  easing: EasingName = "easeInOutCubic",
): number {
  const fn = EASING_FUNCTIONS[easing];
  if (!fn) {
    throw new Error(
      `Unknown easing '${String(easing)}'. Expected one of ${Object.keys(EASING_FUNCTIONS).join(", ")}.`,
    );
  }
  return start + (end - start) * fn(t);
}

File: services/easing_smoothing/easing.py
"""Easing function library shared with the frontend TS port.

Phase 1 — P1-G01.

The function names registered in :data:`EASING_FUNCTIONS` match the contract
enum used in the ``easing_smoothing`` service config
(``linear | easeOutCubic | easeInOutCubic | easeInOutSine``). Snake-case Python
helpers exist for ergonomics, while the lookup map keeps the canonical names so
callers can pass the same string the job manifest carries.
"""

from __future__ import annotations

import math
from typing import Callable
from typing import Final
from typing import Literal


EasingName = Literal[
    "linear",
    "easeOutCubic",
    "easeInOutCubic",
    "easeInOutSine",
]


def _clamp01(value: float) -> float:
    if value != value:  # NaN check without importing math.isnan to keep it cheap
        return 0.0
    if value < 0.0:
        return 0.0
    if value > 1.0:
        return 1.0
    return float(value)


def linear(t: float) -> float:
    return _clamp01(t)


def ease_out_cubic(t: float) -> float:
    x = _clamp01(t)
    inv = 1.0 - x
    return 1.0 - inv * inv * inv


def ease_in_out_cubic(t: float) -> float:
    x = _clamp01(t)
    if x < 0.5:
        return 4.0 * x * x * x
    inv = -2.0 * x + 2.0
    return 1.0 - (inv * inv * inv) / 2.0


def ease_in_out_sine(t: float) -> float:
    x = _clamp01(t)
    return -(math.cos(math.pi * x) - 1.0) / 2.0


EASING_FUNCTIONS: Final[dict[EasingName, Callable[[float], float]]] = {
    "linear": linear,
    "easeOutCubic": ease_out_cubic,
    "easeInOutCubic": ease_in_out_cubic,
    "easeInOutSine": ease_in_out_sine,
}


def interpolate(
    start: float,
    end: float,
    t: float,
    easing: EasingName = "easeInOutCubic",
) -> float:
    """Interpolate between ``start`` and ``end`` using one of the canonical easings.

    ``t`` is clamped to ``[0, 1]`` before easing, matching the TypeScript port
    used by the debug frontend.
    """

    if easing not in EASING_FUNCTIONS:
        raise ValueError(
            f"Unknown easing '{easing}'. Expected one of {sorted(EASING_FUNCTIONS)}."
        )
    eased = EASING_FUNCTIONS[easing](t)
    return start + (end - start) * eased
